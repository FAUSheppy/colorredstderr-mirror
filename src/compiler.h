/*
 * Compiler specific macros.
 *
 * Copyright (C) 2013-2014  Simon Ruderich
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef COMPILER_H
#define COMPILER_H 1

#ifdef HAVE___ATTRIBUTE__
/* Prevent/force inlining. Used to improve performance. */
# define noinline      __attribute__((noinline))
# define always_inline __attribute__((always_inline))
/* Unused parameter. */
# define unused        __attribute__((unused))
/* Mark the function protected, which means it can't be overwritten by other
 * modules (libraries), e.g. with LD_PRELOAD); otherwise same as the default
 * visibility. This causes the compiler not use the PLT (and no relocations)
 * for local calls from inside this module; but the symbols are still
 * exported. This is faster and in this case prevents useless lookups as we
 * hook those functions and nobody else should modify them. Not strictly
 * necessary, but nice to have. */
# define visibility_protected __attribute__((visibility("protected")))
#else
# define noinline
# define always_inline
# define unused
# define visibility_protected
#endif

/* Branch prediction information for the compiler. */
#ifdef HAVE___BUILTIN_EXPECT
# define likely(x)   __builtin_expect(!!(x), 1)
# define unlikely(x) __builtin_expect(!!(x), 0)
#else
# define likely(x)   x
# define unlikely(x) x
#endif

#endif
