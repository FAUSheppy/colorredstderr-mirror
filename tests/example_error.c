/*
 * Test error() and error_at_line(). Non-standard, GNU only.
 *
 * Copyright (C) 2013-2014  Simon Ruderich
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>

#define _GNU_SOURCE /* for program_invocation_name */
#include <errno.h>
#include <error.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

#include "../src/compiler.h"
#include "example.h"


void (*error_print_progname)(void);


static void print_progname(void) {
    fprintf(stderr, "PROG");
}


int main(int argc unused, char **argv unused) {
    pid_t pid;

    char name[] = "./example_error";
    program_invocation_name = name;

    error(0, 0, "<message>");
    error_at_line(0, 0, "file", 42, "<message>");
    FORKED_TEST(pid) { error(1, 0, "<message>"); }
    FORKED_TEST(pid) { error_at_line(1, 0, "file", 42, "<message>"); }

    error(0, ENOMEM, "<message>");
    error_at_line(0, ENOMEM, "file", 42, "<message>");
    error_at_line(0, ENOMEM, "file", 42, "<message>");
    FORKED_TEST(pid) { error(1, ENOMEM, "<message>"); }
    FORKED_TEST(pid) { error_at_line(1, ENOMEM, "file", 42, "<message>"); }

    fflush(stdout);
    printf("\n\n");
    fflush(stdout);

    error_print_progname = print_progname;
    error_one_per_line = 1;

    error(0, 0, "<message>");
    error_at_line(0, 0, "file", 42, "<message>");
    FORKED_TEST(pid) { error(1, 0, "<message>"); }
    FORKED_TEST(pid) { error_at_line(1, 0, "file", 42, "<message>"); }

    error(0, ENOMEM, "<message>");
    error_at_line(0, ENOMEM, "file", 42, "<message>");
    error_at_line(0, ENOMEM, "file", 42, "<message>");
    FORKED_TEST(pid) { error(1, ENOMEM, "<message>"); }

    error_one_per_line = 0;
    FORKED_TEST(pid) { error_at_line(1, ENOMEM, "file", 42, "<message>"); }

    return EXIT_SUCCESS;
}
