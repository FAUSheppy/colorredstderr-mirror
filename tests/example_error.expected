>STDERR>./example_error: <message>
./example_error:file:42: <message>
./example_error: <message>: Cannot allocate memory
./example_error:file:42: <message>: Cannot allocate memory
./example_error:file:42: <message>: Cannot allocate memory
PROG<message>
PROGfile:42: <message>
PROG<message>: Cannot allocate memory
PROGfile:42: <message>: Cannot allocate memory
<STDERR<